rfcs/0050-verification-key-permissions.md

```diff
@@ -0,0 +1,81 @@
+# Verification Key Permissions
+
+This RFC describes the permission scheme for zkApps account verification keys that will be supported at the launch of the Berkeley network.
+
+## Summary
+
+Verification keys control the proofs that can be used when interacting with a zkApp account on-chain. Mina's zkApps allow for specifying permissions when updating various account fields, including the verification key. Updating the verification key is analagous to updating the code of a smart contract on other chains.
+
+In this RFC, we describe a configuration for the verification key permissions that allows developers to make contracts immutable while protecting such contracts from deadlocking upon future upgrades of the protocol.
+
+## Motivation
+
+At the launch of the Berkeley network, the Mina Protocol does not yet guarantee backwards compatability of zkApps in future upgrades. Due to this, it is possible for an immutable contract to break -- any interactions with it that require proofs to be verified against it's verification key will no longer be accepted by the chain. Similarly, the on-chain features that the contract relied on may have been removed or modified.
+
+We wish for zkApps developers to be able to make their contract immutable without putting their contract at risk to breaking upon a future release, rendering any funds locked up in the contract inaccessible.
+
+## Detailed design
+
+NB: This RFC relies on the [protocol versioning RFC](TODO).
+
+In order to prevent a zkApp from breaking upon a future incompatible upgrade of the protocol, we will not allow users to set the `Impossible` or `Proof` permissions on the verification key. Instead, we will add 2 new permissions, `Impossible_during_hard_fork` and `Proof_during_hard_fork`, each of which will carry a specific transaction logic version to represent the hard fork they in which they are valid. Thus, the total set of permissions allowed to be set on verification keys would be:
```

Comment:
Do we need new permissions? In theory, this creates more unreachable states. I would prefer if we simply check the `zkapp_version` in the account and fall back to signature when it's not equal.

Reply:
Hmm... having these explicit permissions is useful in the UX context, as it makes the choice very clear. Perhaps this can be handled only externally, and the internal implementation is exactly how you suggest it. Let me think on this some more and get back to you.